plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/FeatureDevSessionContext.kt

```diff
@@ -49,40 +59,75 @@ class FeatureDevSessionContext(val project: Project) {
     private val gitIgnoreFile = File(projectRoot.path, ".gitignore")
 
     init {
-        ignorePatternsWithGitIgnore = ignorePatterns + parseGitIgnore().map { Regex(it) }
+        ignorePatternsWithGitIgnore = (ignorePatterns + parseGitIgnore().map { Regex(it) }).toList()
     }
 
     fun getProjectZip(): ZipCreationResult {
-        val zippedProject = runReadAction { zipFiles(projectRoot) }
+        val zippedProject = runBlocking { zipFiles(projectRoot) }
         val checkSum256: String = Base64.getEncoder().encodeToString(DigestUtils.sha256(FileInputStream(zippedProject)))
         return ZipCreationResult(zippedProject, checkSum256, zippedProject.length())
     }
 
-    fun ignoreFile(file: File): Boolean = try {
-        ignorePatternsWithGitIgnore.any { p -> p.containsMatchIn(file.path) }
-    } catch (e: Exception) {
-        true
+    private suspend fun ignoreFile(file: File): Boolean = coroutineScope {
+        val deferredResults = ignorePatternsWithGitIgnore.map { pattern ->
+            async {
+                pattern.containsMatchIn(file.path)
+            }
+        }
+        deferredResults.any { it.await() }
     }
 
-    fun ignoreFile(file: VirtualFile): Boolean = ignoreFile(File(file.path))
+    suspend fun ignoreFile(file: VirtualFile): Boolean = ignoreFile(File(file.path))
+
+    suspend fun zipFiles(projectRoot: VirtualFile): File = withContext(Dispatchers.Default) {
+        val files = mutableListOf<VirtualFile>()
+        VfsUtil.visitChildrenRecursively(projectRoot, object : VirtualFileVisitor<Void?>() {
+            override fun visitFile(file: VirtualFile): Boolean {
+                if (file.isFile) {
+                    files.add(file)
+                    return true
+                }
+                return runBlocking {
+                    !ignoreFile(file)
+                }
+            }
+        })
+
+        // Process files in parallel
+        val nonMatchingFiles = mutableListOf<VirtualFile>()
+        coroutineScope {
+            files.map { file ->
+                async {
+                    if (file.isFile && !ignoreFile(file)) {
+                        nonMatchingFiles.add(file)
+                    }
+                }
+            }.awaitAll()
+        }
 
-    private fun zipFiles(projectRoot: VirtualFile): File = createTemporaryZipFile {
-        VfsUtil.collectChildrenRecursively(projectRoot).map { virtualFile -> File(virtualFile.path) }.forEach { file ->
-            if (file.isFile() && !ignoreFile(file)) {
+        createTemporaryZipFileAsync { zipOutput ->
```

Comment:
also don't see the point of this being async if we have to wait for the file to be completely created before it gets returned

it would make more sense to do an s3 streaming upload, which would be a more involved change

Reply:
sure